File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int ans =0;
        int n = s.length();
        Map<Character,Integer>track = new HashMap<>();
        int left = 0; 
        for(int right = 0 ; right < n ; right++){
            char ch = s.charAt(right);
            // Only check if character exists in map
            if (track.containsKey(ch) && track.get(ch) >= left) {
                left = track.get(ch) + 1;
            }

            track.put(ch , right);//puting 
            ans  = Math.max(ans , right -left +1); 
        }

        return ans;
        
    }
}

//try doing with loop 
// then use sliding windows 
//when ever get again then left = ind +1 and replace also in the map 
//left is import if we move ahead and we are not removing the present value thaswhy 
// abba -- for , why we use left , why don't empty 

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int maxlength = 0; // Final answer to store the maximum length
        vector<int> ans(256, -1); // To store the last index of each character
        int n = s.length();

        int left = 0; // Start of the sliding window
        for (int right = 0; right < n; right++) { // End of the sliding window
            // If the character is already in the current window
            if (ans[s[right]] >= left) {
                left = ans[s[right]] + 1; // Move the left pointer past the duplicate
            }

            ans[s[right]] = right; // Update the last seen index of the character
            maxlength = max(maxlength, right - left + 1); // Update the maximum length
        }

        return maxlength;
    }
};
